pkg/controllermanager/controller/shoot/shoot_maintenance_control.go

```diff
@@ -224,3 +219,81 @@ func hasMaintainNowAnnotation(shoot *gardenv1beta1.Shoot) bool {
+func MaintainMachineImage(shoot *gardenv1beta1.Shoot, cloudProfile *gardenv1beta1.CloudProfile, shootCurrentImage *gardenv1beta1.MachineImage) (bool, *gardenv1beta1.MachineImage, error) {
+	machineImagesFound, machineImages, err := helper.DetermineMachineImages(*cloudProfile, shootCurrentImage.Name)
+	if err != nil {
+		return false, nil, fmt.Errorf("failure while determining the machine images in the CloudProfile: %s", err.Error())
+	}
+	if !machineImagesFound || len(machineImages) == 0 {
+		return false, nil, fmt.Errorf("failure while determining the machine images in the CloudProfile: no machineImage with name '%s' (specified in shoot) could be found in the cloud cloudProfile '%s'", shootCurrentImage.Name, cloudProfile.Name)
+	}
+
+	// checks if forceUpdate Time & applicable (has the specified version)
+	forceUpdateRequired, forceUpdateTime := ForceUpdateRequired(shootCurrentImage, machineImages)
+
+	// safely dereference due to defaulting in api server
+	autoUpdateOfMachineImageConfigured := *shoot.Spec.Maintenance.AutoUpdate.MachineImageVersion
+	if forceUpdateRequired || autoUpdateOfMachineImageConfigured {
+		versionUpdatePossible, updateMachineImage, err := VersionUpdatePossible(shootCurrentImage, machineImages)
+		if err != nil {
+			return false, nil, nil
+		}
+		if forceUpdateRequired && !versionUpdatePossible {
+			return false, nil, fmt.Errorf("Cannot force update machine Image of shoot '%s'. A consecutive version could not be determined - the shoot is already on the latest image version configured in the cloud cloudProfile. Shoot image: '%s/%s'. ForceUpdateTime of this image version as defined in CloudProfile: '%s'. ", shoot.Name, shootCurrentImage.Name, shootCurrentImage.Version, *forceUpdateTime)
+		}
+		if versionUpdatePossible {
+			return true, updateMachineImage, nil
+		}
+	}
+	return false, nil, nil
+}
+
+// VersionUpdatePossible checks if the shoot's current image can be updated to a higher semantic version
+func VersionUpdatePossible(shootCurrentImage *gardenv1beta1.MachineImage, machineImages []gardenv1beta1.MachineImage) (bool, *gardenv1beta1.MachineImage, error) {
+	latestSemVerVersion, latestMachineImage, err := helper.DetermineLatestMachineImage(machineImages)
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to determine latest machine image in cloud profile: %s", err.Error())
+	}
+
+	shootImageSemVerVersion, err := semver.NewVersion(shootCurrentImage.Version)
+	if err != nil {
+		return false, nil, fmt.Errorf("error while parsing machine image version of shoot (not valid): %s", err.Error())
+	}
+
+	if latestSemVerVersion.GreaterThan(shootImageSemVerVersion) {
+		return true, &latestMachineImage, nil
+	}
+	return false, nil, nil
+}
+
+// ForceUpdateRequired checks if the shoots current machine image has to be forcefully updated
+func ForceUpdateRequired(shootCurrentImage *gardenv1beta1.MachineImage, imagesCloudProvider []gardenv1beta1.MachineImage) (bool, *string) {
```

Comment:
Only return `bool` if timestamp is not required.

Reply:
Could you elaborate? Thanks!

Reply:
It's connected to my previous comment. If you get rid of the `forceUpdateRequired && !versionUpdatePossible` case, you won't need the `ForceUpdateTime`.